feat: support switch-case replacements and definitions at root level

TTtie opened the pull request:
Adds support for switch-case style replacement in message actions, inspired by pattern matching found in modern programming languages, for example (preliminary syntax, some details may be subject to change):

```xml
<message text="The outcome is {outcome}">
    <replacements>
        <!-- 
             value supports formulas and may be omitted 
             (in this case, the match attribute in <case>
             is disallowed)
        -->
        <switch id="outcome" value="variable">
            <!-- matches if the formula result is 0 -->
            <case match="0" result="Outcome 1"/>
            
            <!-- matches if the formula result is 1 AND only-team-1 allows -->
            <case match="1" filter="only-team-1" result="`aOutcome 2"/>
            
            <!-- matches if only-team-2 allows regardless of formula result -->
            <case filter="only-team-2" result="`bOutcome 3"/>
            
            <!--
                full syntax (matches if the formula result is between 2 and 5,
                and the inner filter allows)
            -->
            <case>
                <match>2..5</match>
                <filter>
                    <all id="not-moving">
                        <variable var="player.vel_x">0</variable>
                        <variable var="player.vel_y">..0</variable>
                        <variable var="player.vel_z">0</variable>
                    </all>
                </filter>
                <result>`cOutcome 4</result>
            </case>
            
            <!-- a fallback value if no cases match (defaults to an empty component) -->
            <fallback>`dA fallback outcome</fallback>
        </switch>
    </replacements>
</message>
```

Additionally, replacements can now be specified at the root level:
```xml
<map>
    <replacements>
        <!--
            scope is required for all replacements except <player />
            and it will be enforced if specified
        -->
        <decimal id="decimal" value="money" scope="player" format="0.00"/>
        <switch id="team" scope="team">
            <case filter="only-team-1" result="`aLime"/>
        </switch>
        
        <!-- 
            This will error as only-team-1 is team-scoped
            Not true as of commit 54666274 anymore. `only-team-1` will match as it abstains.
        -->
        <switch id="team2" scope="match">
            <case filter="only-team-1" result="`aLime"/>
        </switch>
    </replacements>
    <actions>
        <action id="action" scope="match">
            <switch-scope inner="player">
                <message text="The variable is {variable} and your team is {team}!">
                    <replacements>
                        <!--
                            the syntax for replacement references is
                            <replacement id="local-id">global-id</replacement>
                        -->
                        <replacement id="variable">decimal</replacement>
                        <replacement id="team">team</replacement>
                    </replacements>
                </message>
            </switch-scope>
        </action>
    </actions>
</map>
```

---

Launch checklist:
- [x] Cannot use higher-scope filters in `<case>` than defined in `<switch>`
- [x] Cannot specify higher-scope replacements than defined in `<action>`
- [x] Cannot use lower-scope variables in `<switch>` value
- [x] Validation???
  - [x] Using a reference of lower scope replacement should not work in a higher-scope action
  - [x] Declaring a lower scope replacement in a higher scope action should not work?

TTtie committed:
feat(replacements): add switch-case replacement

Adds support for switch-case style replacement in message actions, inspired by pattern matching found in modern programming languages, for example

```xml
<action filter="game=1"><message text="`3Up Next: `b[1] Hungry Reindeer"/></action>
<action filter="game=2"><message text="`3Up Next: `b[2] Present Hunt"/></action>
<action filter="game=3"><message text="`3Up Next: `b[3] Winter Speedway"/></action>
<action filter="game=4"><message text="`3Up Next: `b[4] Parkour"/></action>
<action filter="game=5"><message text="`3Up Next: `b[5] Chimney Sweepers"/></action>
<action filter="game=6"><message text="`3Up Next: `b[6] Frosted Corridors"/></action>
```
becomes
```xml
<message text="`3Up Next: `b[{game}] {game-name}">
    <replacements>
        <decimal id="game" value="game"/>
        <switch id="game-name" value="game">
            <case value="1" result="Hungry Reindeer"/>
            <case value="2" result="Present Hunt"/>
            <case value="3" result="Winter Speedway"/>
            <case value="4" result="Parkour"/>
            <case value="5" result="Chimney Sweepers"/>
            <case value="6" result="Frosted Corridors"/>
        </switch>
    </replacements>
</message>
```

It is also possible to use them with filters:
```xml
<message text="The bomb has been planted on {bombsite} by {holder}">
    <replacements>
        <player id="holder" var="bomb_holder" fallback="an unknown player"/>
        <switch id="bombsite">
            <case filter="planted-a" result="`c${bombsite-a-name}"/>
            <case filter="planted-b" result="`c${bombsite-b-name}"/>
        </switch>
    </replacements>
</message>
```

and also to mix them together:
```xml
<message text="Your score is {score} ({score_num})!">
    <replacements>
        <decimal id="score_num" value="score"/>
        <switch id="score" value="score">
            <case value="0..70" filter="only-attackers" result="`cvery low :("/>
            <case value="0..70" filter="only-defenders" result="`9very low :(" />
            <case value="70..100" filter="only-attackers" result="`clow :|"/>
            <case value="70..100" filter="only-defenders" result="`9low :|" />
            <case value="100..130" filter="only-attackers" result="`chigh :)"/>
            <case value="100..130" filter="only-defenders" result="`9high :)" />
            <fallback>off the charts :O</fallback>
        </switch>
    </replacements>
</message>
```

Signed-off-by: TTtie <me@tttie.cz>

Pablete1234 commented on core/src/main/java/tc/oc/pgm/action/replacements/Replacement.java:
```diff
@@ -0,0 +1,27 @@
+   * @param audience The audience to use when replacing
+   * @return The replacement component
+   */
+  ComponentLike replace(S audience);
```
You're not *replacing* you're creating a component, so i feel calling this method replace is a bit misleading
Maybe `create` or `get` would fit better

Pablete1234 commented on core/src/main/java/tc/oc/pgm/action/replacements/Replacement.java:
```diff
@@ -0,0 +1,27 @@
+   *
+   * @return The scope of the replacement
+   */
+  Class<S> getScope();
```
You can likely get away with not having a scope, but rather a *checking* function, since some replacements dont need any specific scope.

Think of it in the same spirit as `respondsTo` in filters.

Reason behind this is that if a replacement works for `Match` it should also work for party or player, there's no reason it shouldn't, at least not with the current replacements. If in the future a new replacement requires something else you can still implement that replacement as more strict, but for the ones we have it should be more lenient

Pablete1234 commented on core/src/main/java/tc/oc/pgm/action/replacements/ReplacementParser.java:
```diff
@@ -0,0 +1,193 @@
+      throws InvalidXMLException {
+    scope = parseScope(el, scope);
+    Formula<T> formula = parser.formula(scope, el, "value").required();
+    var formatString = parser.string(el, "format").attr().orNull();
```
`.optional().map(DecimalFormat::new).orElse(DEFAULT_FORMAT);`

Pablete1234 commented on core/src/main/java/tc/oc/pgm/action/replacements/ReplacementParser.java:
```diff
@@ -0,0 +1,193 @@
+    var nameStyle = parser.parseEnum(NameStyle.class, el, "style").optional(NameStyle.VERBOSE);
+    // Double cast is required, otherwise the compiler complains about type incompatibility
+    @SuppressWarnings("unchecked,RedundantCast")
+    Class<T> _scope = scope == null ? (Class<T>) (Object) Filterable.class : scope;
```
You're doing the most scope-tricks for the replacement that needs them the least. The player replacement can literally work with any scope, it shouldnt even need to do any figuring-out

Pablete1234 commented on core/src/main/java/tc/oc/pgm/action/replacements/ReplacementParser.java:
```diff
@@ -0,0 +1,193 @@
+
+      var result = parser.formattedText(innerEl, "result").required();
+
+      if (filter != null) {
```
This isn't the place to write this code, make this part of the filter parser so you can `parser.filter(el, "filter").respondsTo(scope).orNull()`

Pablete1234 commented on core/src/main/java/tc/oc/pgm/action/replacements/ReplacementParser.java:
```diff
@@ -0,0 +1,193 @@
+      branches.add(new SwitchReplacement.Branch(
+          result,
+          valueRange == null ? Range.all() : valueRange,
+          filter == null ? StaticFilter.ALLOW : filter));
```
You can use `orAllow()` at the end of the filter parser to avoid the nullcheck

Pablete1234 commented on core/src/main/java/tc/oc/pgm/action/replacements/Replacement.java:
```diff
@@ -0,0 +1,27 @@
+ *
+ * @param <S> The scope of the replacement
+ */
+public interface Replacement<S extends Audience> extends FeatureDefinition {
```
scopes should be `extends Filterable<?>`

Pablete1234 commented on core/src/main/java/tc/oc/pgm/action/replacements/ReplacementParser.java:
```diff
@@ -0,0 +1,193 @@
+    return replacement;
+  }
+
+  private void validateScope(Replacement<?> definition, Class<?> scope, Element el)
+      throws InvalidXMLException {
+    Class<?> definitionScope = definition.getScope();
+    if (!definitionScope.isAssignableFrom(scope))
+      throw new InvalidXMLException(
+          "Wrong replacement scope, got "
+              + definitionScope.getSimpleName()
+              + " but expected "
+              + scope.getSimpleName(),
+          el);
+  }
```
As mentioned before, relegate this to a method within the replacement; let it decide what the logic for rejecting a scope should be

calcastor commented:
Testing with changes from [this repo](https://github.com/TTtie/occ-community-maps/commits/refactor/use-switch-case-replacements/), with the exception of Grand Dueling and Chromatic Conquest which differ too much upstream for me to rebase the changes right now, everything seems to work as intended.

TTtie committed:
refactor/fix: move validation errors into replacements, verify scope when parsing inline replacement

Logical correctness is not guaranteed.

Signed-off-by: TTtie <me@tttie.cz>

TTtie committed:
lint

Signed-off-by: TTtie <me@tttie.cz>

Pablete1234 commented on core/src/main/java/tc/oc/pgm/util/xml/parsers/FilterBuilder.java:
```diff
@@ -27,6 +29,26 @@ public FilterBuilder dynamic(Class<? extends Filterable<?>> type) {
+                    + ". The filter "
+                    + definition.getDefinitionType().getSimpleName()
+                    + " only responds to "
+                    + Filterables.scope(definition).getSimpleName()
```
This can and will throw for filters that do not belong to any scope, which is possible in non-dynamic filters.

TTtie committed:
feat: ignore "match" attribute when switch value is unspecified

Instead of throwing a parsing error, the parser will now emit an unused node warning.

Signed-off-by: TTtie <me@tttie.cz>

TTtie committed:
refactor!: remove FilterBuilder#respondsTo

The usage of Filter#respondsTo hints that it is meant to be used with dynamic filters, which are covered by FilterBuilder#dynamic.

This affects the behavior of the switch-case replacement by allowing any filter to be passed. This also means that filters that abstain will match if used in a case branch.

Signed-off-by: TTtie <me@tttie.cz>

Pablete1234 commented on core/src/main/java/tc/oc/pgm/action/replacements/ReplacementParser.java:
```diff
@@ -0,0 +1,153 @@
+
+    for (var innerEl : children) {
+      var valueRange = formula != null ? parser.doubleRange(innerEl, "match").orNull() : null;
+      var filter = parser.filter(innerEl, "filter").optional(() -> {
+        if (valueRange == null)
+          throw new InvalidXMLException(
+              "At least a filter or a match attribute must be specified", innerEl);
+        return StaticFilter.ALLOW;
+      });
```
If you don't specify a value, and do specify a match, you could end up with a weird error message:

"Unused node 'match'" alongside an "At least a filter or a match..." which may leave you puzzled if you were a mapdev

maybe an option would be:
```suggestion
      var filterParse = parser.filter(innerEl, "filter");
      var filter = valueRange == null ? filterParse.required() : filterParse.orAllow();
```
which will just mark filter as required (without mentioning match at all) if you've not specified match.
not sure if it's really better or worse, maybe the err message should be different depending on if value is specified or not? i really don't know what's better

TTtie replied:
> If you don't specify a value, and do specify a match, you could end up with a weird error message:
> 
> "Unused node 'match'" alongside an "At least a filter or a match..." which may leave you puzzled if you were a mapdev

That will never happen. Parsing errors interrupt the whole map loading process, and unused node warnings are the very last step in that process. The error message probably needs a rewrite given that `match="..."` is ignored when there's no value though.

TTtie committed:
fix: provide a more helpful error if filter is omitted and <switch> match is also unspecified

Signed-off-by: TTtie <me@tttie.cz>

TTtie committed:
chore: remove the word attribute from error messages

Child elements are also allowed, and in that case, the error wouldn't make sense

Signed-off-by: TTtie <me@tttie.cz>
